File: internal/wailsapi/dto.go
package wailsapi

import (
	"time"

	"github.com/SellswordSoftware/justbookmarks/internal/bookmarks"
)

type BookmarkDTO struct {
	ID           string `json:"id"`
	Title        string `json:"title"`
	URL          string `json:"url"`
	Icon         string `json:"icon"`
	IconURI      string `json:"iconURI"`
	AddDate      string `json:"addDate"`
	LastModified string `json:"lastModified"`
	Meta         string `json:"meta"`
}

type BookmarkCreateDTO struct {
	Title   string `json:"title"`
	URL     string `json:"url"`
	Icon    string `json:"icon"`
	IconURI string `json:"iconURI"`
	Meta    string `json:"meta"`
}

type BookmarkPatchDTO struct {
	Title   *string `json:"title,omitempty"`
	URL     *string `json:"url,omitempty"`
	Icon    *string `json:"icon,omitempty"`
	IconURI *string `json:"iconURI,omitempty"`
	Meta    *string `json:"meta,omitempty"`
}

type FolderDTO struct {
	ID           string    `json:"id"`
	Name         string    `json:"name"`
	Icon         string    `json:"icon"`
	AddDate      string    `json:"addDate"`
	LastModified string    `json:"lastModified"`
	Meta         string    `json:"meta"`
	Children     []NodeDTO `json:"children"`
}

type NodeDTO struct {
	ID       string             `json:"id"`
	Type     bookmarks.NodeType `json:"type"`
	Folder   *FolderDTO         `json:"folder"`
	Bookmark *BookmarkDTO       `json:"bookmark"`
}

type MoveResult struct {
	MovedNodes  []bookmarks.FlatNode `json:"movedNodes"`
	OldParentID string               `json:"oldParentId"`
	NewParentID string               `json:"newParentId"`
	NewIndex    int                  `json:"newIndex"`
}

func toNodeDTOs(nodes []bookmarks.Node) []NodeDTO {
	result := make([]NodeDTO, 0, len(nodes))
	for _, node := range nodes {
		result = append(result, toNodeDTO(node))
	}
	return result
}

func toNodeDTO(node bookmarks.Node) NodeDTO {
	dto := NodeDTO{
		ID:   node.ID(),
		Type: node.Type,
	}

	if node.Folder != nil {
		dto.Folder = &FolderDTO{
			ID:           node.Folder.ID,
			Name:         node.Folder.Name,
			Icon:         node.Folder.Icon,
			AddDate:      formatTimestamp(node.Folder.AddDate),
			LastModified: formatTimestamp(node.Folder.LastModified),
			Meta:         node.Folder.Meta,
			Children:     toNodeDTOs(node.Folder.Children),
		}
	}

	if node.Bookmark != nil {
		dto.Bookmark = &BookmarkDTO{
			ID:           node.Bookmark.ID,
			Title:        node.Bookmark.Title,
			URL:          node.Bookmark.URL,
			Icon:         node.Bookmark.Icon,
			IconURI:      node.Bookmark.IconURI,
			AddDate:      formatTimestamp(node.Bookmark.AddDate),
			LastModified: formatTimestamp(node.Bookmark.LastModified),
			Meta:         node.Bookmark.Meta,
		}
	}

	return dto
}

func toBookmarkCreate(dto BookmarkCreateDTO) bookmarks.Bookmark {
	return bookmarks.Bookmark{
		Title:   dto.Title,
		URL:     dto.URL,
		Icon:    dto.Icon,
		IconURI: dto.IconURI,
		Meta:    dto.Meta,
	}
}

func toBookmarkPatch(dto BookmarkPatchDTO) bookmarks.BookmarkPatch {
	return bookmarks.BookmarkPatch{
		Title:   dto.Title,
		URL:     dto.URL,
		Icon:    dto.Icon,
		IconURI: dto.IconURI,
		Meta:    dto.Meta,
	}
}

func formatTimestamp(value time.Time) string {
	if value.IsZero() {
		return ""
	}
	return value.UTC().Format(time.RFC3339)
}

File: src-tauri/src/dto.rs
use serde::{Deserialize, Serialize};

use crate::models::{Bookmark, BookmarkPatch, FlatNode, Node};

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BookmarkCreateDto {
    pub title: String,
    pub url: String,
    pub icon: String,
    #[serde(rename = "iconURI")]
    pub icon_uri: String,
    pub meta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BookmarkDto {
    pub id: String,
    pub title: String,
    pub url: String,
    pub icon: String,
    #[serde(rename = "iconURI")]
    pub icon_uri: String,
    pub add_date: String,
    pub last_modified: String,
    pub meta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FolderDto {
    pub id: String,
    pub name: String,
    pub icon: String,
    pub add_date: String,
    pub last_modified: String,
    pub meta: String,
    pub children: Vec<NodeDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NodeDto {
    pub id: String,
    #[serde(rename = "type")]
    pub node_type: u8,
    pub folder: Option<FolderDto>,
    pub bookmark: Option<BookmarkDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MoveResult {
    pub moved_nodes: Vec<FlatNode>,
    pub old_parent_id: String,
    pub new_parent_id: String,
    pub new_index: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TreeStats {
    pub folders: i32,
    pub bookmarks: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FolderMergeItem {
    pub path: String,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BookmarkMergeItem {
    pub folder_path: String,
    pub title: String,
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BookmarkConflictItem {
    pub folder_path: String,
    pub existing_title: String,
    pub incoming_title: String,
    pub url: String,
    pub existing_meta: String,
    pub incoming_meta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MergePreview {
    pub folders_to_add: Vec<FolderMergeItem>,
    pub bookmarks_to_add: Vec<BookmarkMergeItem>,
    pub duplicate_bookmarks: Vec<BookmarkMergeItem>,
    pub potential_updates: Vec<BookmarkConflictItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MergeApplyResult {
    pub folders_added: i32,
    pub bookmarks_added: i32,
    pub duplicates_skipped: i32,
    pub potential_updates: i32,
}

impl From<BookmarkCreateDto> for Bookmark {
    fn from(value: BookmarkCreateDto) -> Self {
        Self {
            title: value.title,
            url: value.url,
            icon: value.icon,
            icon_uri: value.icon_uri,
            meta: value.meta,
            ..Bookmark::default()
        }
    }
}

impl From<BookmarkPatch> for BookmarkDto {
    fn from(value: BookmarkPatch) -> Self {
        Self {
            title: value.title.unwrap_or_default(),
            url: value.url.unwrap_or_default(),
            icon: value.icon.unwrap_or_default(),
            icon_uri: value.icon_uri.unwrap_or_default(),
            meta: value.meta.unwrap_or_default(),
            ..Self::default()
        }
    }
}

impl From<Bookmark> for BookmarkDto {
    fn from(value: Bookmark) -> Self {
        Self {
            id: value.id,
            title: value.title,
            url: value.url,
            icon: value.icon,
            icon_uri: value.icon_uri,
            add_date: value.add_date,
            last_modified: value.last_modified,
            meta: value.meta,
        }
    }
}

impl From<Node> for NodeDto {
    fn from(value: Node) -> Self {
        Self {
            id: value.id,
            node_type: value.node_type,
            folder: value.folder.map(|folder| FolderDto {
                id: folder.id,
                name: folder.name,
                icon: folder.icon,
                add_date: folder.add_date,
                last_modified: folder.last_modified,
                meta: folder.meta,
                children: folder.children.into_iter().map(NodeDto::from).collect(),
            }),
            bookmark: value.bookmark.map(BookmarkDto::from),
        }
    }
}
